[1.19.X] Adding more precise events for Advancements

28Pollux28 opened the pull request:
Replace the old AdvancementEvent that was only firing when a player was completing an advancement by two new Events :
- ~~AdvancementEvent.AdvancementEarnEvent~~ AdvancementEvents.AdvancementEarnEvent : Same as the old AdvancementEvent, fires when a Player completes an advancement.
- ~~AdvancementEvent.AdvancementProgressEvent~~ AdvancementEvents.AdvancementProgressEvent : New Event that fires when a Player progresses an advancement's criterion. It ships the player, the advancement that was progressed, the AdvancementProgress for the Advancement, the criterion String that was progressed and the type of progress (~~AdvancementEvent.ProgressType~~AdvancementEvents.AdvancementProgressEvent.ProgressType, either GRANT or REVOKE)
- Introduces a ProgressType enum for CriterionProgress, with values GRANT and REVOKE

28Pollux28 committed:
added AdvancementEvent.AdvancementEarnEvent
added AdvancementEvent.AdvancementProgressEvent
added AdvancementEvent.ProgressType
added ForgeEventFactory.onAdvancementEarnedEvent
added ForgeEventFactory.onAdvancementProgressedEvent
removed ForgeHooks.onAdvancement

diesieben07 reviewed (changes requested):
- Please review the code style. You have a bunch of braces on the same line as well as wrong/missing whitespace
- This is not backwards compatible. Mods that listen to the old event will now receive both new events, which is not what they expect.
- You have removed methods, which is a binary break and thus can only be done on a Minecraft version change.

28Pollux28 committed:
added backport compatibility
marked previous event deprecated and added todo for removal in 1.19
moved new Events from AdvancementEvent to AdvancementEvents
fixed code style

sciwhiz12 commented on src/test/resources/META-INF/mods.toml:
```diff
@@ -174,5 +174,6 @@ license="LGPL v2.1"
 [[mods]]
     modId="capabilities_test"
-
+[[mods]]
+    modId="advancement_event"
 # ADD ABOVE THIS LINE
```
Please keep the blank line above the comment.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,58 @@
+        String criterion = event.criterion;
+        AdvancementEvents.AdvancementProgressEvent.ProgressType progressType = event.progressType;
+        String action;
+        if(progressType == ProgressType.GRANT)
```
Formatting: space before parentheses in `if`-expression:
```suggestion
        if (progressType == ProgressType.GRANT)
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,58 @@
+package net.minecraftforge.debug.entity.player;
```
Missing license header, please run the `updateLicenses`/`licenseFormat` task.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,58 @@
+{
+    public static final String MOD_ID = "advancement_event";
+    public static final boolean ENABLED = false;
+    public static final Logger LOGGER = LogManager.getLogger();
```
The logger field should be `private`.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,58 @@
+    public AdvancementEventTest()
+    {
+        IEventBus forgeBus = MinecraftForge.EVENT_BUS;
+        if(ENABLED)
```
Formatting: space before parentheses in `if`-expression:
```suggestion
        if (ENABLED)
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,58 @@
+
+    public AdvancementEventTest()
+    {
+        IEventBus forgeBus = MinecraftForge.EVENT_BUS;
```
For the sake of compactness, this should be inlined in the two calls below.

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvents.java:
```diff
@@ -0,0 +1,97 @@
+        }
+
+        private final Advancement advancement;
+        public final AdvancementProgress advancementProgress;
+        public final String criterion;
+        public final ProgressType progressType;
```
Encapsulate these fields with a public getter and mark them `private`.

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvents.java:
```diff
@@ -0,0 +1,97 @@
+    {
+        public enum ProgressType
+        {
+            GRANT,REVOKE
```
Formatting: space after comma between enum value declarations:
```suggestion
            GRANT, REVOKE
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvents.java:
```diff
@@ -0,0 +1,97 @@
+        public final String criterion;
+        public final ProgressType progressType;
+
+        /**
+         *
+         * @param player The Event's {@link net.minecraft.world.entity.player.Player}
+         * @param progressed The {@link Advancement} that was progressed
+         * @param advancementProgress The {@link net.minecraft.advancements.AdvancementProgress} of the {@link net.minecraft.advancements.Advancement}
+         * @param criterion The {@link String} representing the {@link net.minecraft.advancements.CriterionProgress} that was progressed.
+         * @param progressType Defines what {@link net.minecraftforge.event.entity.player.AdvancementEvents.AdvancementProgressEvent.ProgressType} was the {@link net.minecraft.advancements.CriterionProgress}
+         */
```
Please avoid the use of `{@link}` tags in `@param` tags when referencing the type of the parameter (or other types already referenced in the method javadocs or declaration.
```suggestion
        /**
         * @param player the player which progressed the advancement's criterion
         * @param progressed the advancement which was progressed
         * @param advancementProgress the progress of the advancement
         * @param criterion the name of the criterion which was progressed
         * @param progressType the type of progress for the advancement criterion
         */
```

Alternatively, delete this javadoc on the constructor and move the javadocs to the getters for the fields:
```java
        /**
         * {@return the advancement which was progressed}
         */
        public Advancement getAdvancement()
        {
            return advancement;
        }

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvents.java:
```diff
@@ -0,0 +1,97 @@
+import net.minecraft.world.entity.player.Player;
+import net.minecraftforge.common.MinecraftForge;
+
+public class AdvancementEvents
```
This style of a holder class for events that itself is not an event is discouraged. Please move these events to be inner/nested classes of `AdvancementEvent`. You should also change the events then to subclass `AdvancementEvent` (we are still in the period of breaking changes, so this change of API is allowed to my knowledge).

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -32,6 +34,7 @@
  * <br>
  * This event is fired on the {@link MinecraftForge#EVENT_BUS}.
  */
+@Deprecated()//todo : remove and use AdvancementEvents instead in 1.19
```
This class should not be deprecated for removal, and the events classes in `AdvancementEvents` should be moved to be nested/inner classes of this class instead.

sciwhiz12 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -976,6 +976,10 @@ public static CriticalHitEvent getCriticalHit(Player player, Entity target, bool
+    /**
+     * @deprecated See {@linkplain ForgeEventFactory}
+     */
+    @Deprecated //todo: remove and use ForgeEventFactory.onAdvancementEarned instead in 1.19
```
Mark deprecations for removal as such in the annotation:
```suggestion
    @Deprecated(forRemoval = true, since = "1.18.1") // TODO: remove and use ForgeEventFactory.onAdvancementEarned instead in 1.19
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvents.java:
```diff
@@ -0,0 +1,97 @@
+
+        private final Advancement advancement;
+        public final AdvancementProgress advancementProgress;
+        public final String criterion;
```
As this is not the criterion progress itself but the name of the criterion, this should be named as `criterionName`.

sciwhiz12 commented on patches/minecraft/net/minecraft/server/PlayerAdvancements.java.patch:
```diff
@@ -20,11 +20,26 @@
                 this.f_135962_.m_11264_(new TranslatableComponent("chat.type.advancement." + p_135989_.m_138320_().m_14992_().m_15548_(), this.f_135968_.m_5446_(), p_135989_.m_138330_()), ChatType.SYSTEM, Util.f_137441_);
              }
-+            net.minecraftforge.common.ForgeHooks.onAdvancement(this.f_135968_, p_135989_);
++            net.minecraftforge.common.ForgeHooks.onAdvancement(this.f_135968_,p_135989_);//todo:remove in 1.19
```
Formatting: space after comma in method parameters list, and some additional spaces around and in the comment:
```suggestion
+            net.minecraftforge.common.ForgeHooks.onAdvancement(this.f_135968_, p_135989_); // todo: remove in 1.19
```

28Pollux28 committed:
Implement changes suggested by sciwhiz12
Fix mods.toml
Add License Header
Fix visibility of LOGGER
Inlined calls
Fix formatting
Corrected Deprecated annotations
Moved back new events to AdvancementEvent

sciwhiz12 reviewed (changes requested):
The introduction of the two new events as subclasses of `AdvancementEvent` in this PR poses a problem now, as this breaks the current contract of `AdvancementEvent` and we are too close to an RB to risk the breaking change. 

Therefore, the two new events -- `AdvancementEarnEvent` and `AdvancementProgressEvent` -- need to not extend `AdvancementEvent` for now, with that change slated to be done in 1.19 instead (when breaking changes are allowed):
 - Both new events should extend `PlayerEvent`, with a TODO and note to change the event in 1.19 to extend `AdvancementEvent`;
 - Because of the above, both events must now have an `advancement` field and `getAdvancement()` getter.
 - `AdvancementEvent` itself should be deprecated (but _not_ for removal), with a `@deprecated` note in its javadoc that current users of the event should move to the two new events instead.
 - Keep the hook and calling place for `AdvancementEvent` with its current TODO to remove in 1.19.

---
After updating your PR, please remember to run the `updateLicenses`/`licenseFormat` task to update the license headers in your test mod to the new format.

sciwhiz12 commented:
Hello! @28Pollux28, please leave a comment whether you wish to continue this PR and update it to the latest active version of 1.19.2 (or leave an explanation why this PR's changes are not applicable to 1.19.2). If we do not hear back from you after some time, this PR will be closed for inactivity.

If anyone else wishes to continue the work of this PR, please create a PR with a link to this PR for reference.
Thank you for your contributions. 👋

28Pollux28 commented:
@sciwhiz12 Tbh I had a really bad experience with the whole forge community, either from users or from doing PRs. I proposed changes, had to do some changes, then was asked to revert it back, then to keep the changes and then to revert it back. At that point it seemed to me that proposing such a small feature and going through all of this wasn't worth it and just implemented my version in my own code.
Now I understand it would be a great addition to the forge ecosystem so I might try and make an update for 1.19.2, but I won't go to the same length as before, too much trouble and I'm out.

28Pollux28 committed:
Update to 1.19.x

28Pollux28 commented:
Branch is still named 1.18, but is based on 1.19 code

sciwhiz12 commented on src/test/resources/META-INF/mods.toml:
```diff
@@ -14,7 +14,8 @@ license="LGPL v2.1"
     modId="fluid_type_test"
 [[mods]]
     modId="potentialspawnsevent_test"
-
```
Please keep this blank line.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,75 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.debug.entity.player;
```
Outdated license header, please run the `updateLicenses`/`licenseFormat` task.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,75 @@
+import net.minecraftforge.fml.common.Mod;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
```
Please use SLF4J's `Logger` and `LogUtils.getLogger()`.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,75 @@
+        {
+            action = "revoked";
+        }
+        LOGGER.info("Player {} progressed advancement {}. It was {} progress on {} criterionName. It has completed {}% of the achievement", player, advancement.getId(), action, criterionName, advancementProgress.getPercent()*100);
```
In the interests of readability, please split this line into two by the end of the message:
```suggestion
        LOGGER.info("Player {} progressed advancement {}. They were {} progress on {}. They have completed {}% of the achievement", 
        		player, advancement.getId(), action, criterionName, advancementProgress.getPercent() * 100);
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,75 @@
+        Player player = event.getEntity();
+        LOGGER.info("Player {} earned advancement {} and was awarded {}", player, advancement.getId(), advancement.getRewards().toString());
+    }
+    public void onAdvancementProgressEvent(AdvancementProgressEvent event)
```
Please place a blank line between these methods.
```suggestion
    }

    public void onAdvancementProgressEvent(AdvancementProgressEvent event)
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/AdvancementEventTest.java:
```diff
@@ -0,0 +1,75 @@
+public class AdvancementEventTest
+{
+    public static final String MOD_ID = "advancement_event_test";
+    public static final boolean ENABLED = false;
```
These should be `private`.

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -32,4 +34,99 @@ public Advancement getAdvancement()
+
+        /**
+         *
+         * @return The advancement that was progressed.
```
```suggestion
         * {@return the advancement that was progressed}
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -32,4 +34,99 @@ public Advancement getAdvancement()
+        }
+
+        /**
+         * {@return The AdvancementProgress of the Advancement}
```
```suggestion
         * {@return the progress of the advancement}
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -32,4 +34,99 @@ public Advancement getAdvancement()
+        }
+
+        /**
+         * {@return The Criterion's name that was progressed}
```
```suggestion
         * {@return the name of the criterion that was progressed}
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -32,4 +34,99 @@ public Advancement getAdvancement()
+        }
+
+        /**
+         * {@return The ProgressType of the progressed Advancement}
```
```suggestion
         * {@return the type of progress for the criterion in this event}
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -32,4 +34,99 @@ public Advancement getAdvancement()
+        public enum ProgressType
+        {
+            GRANT, REVOKE
+        }
```
Please move this to the end of the class, so all fields are grouped together.

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -32,4 +34,99 @@ public Advancement getAdvancement()
+
+        /**
+         *
+         * @return The advancement that was earned.
```
```suggestion
         * {@return the advancement that was earned}
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -17,6 +18,7 @@
  * This event is fired on the {@link MinecraftForge#EVENT_BUS}.
+ * @Deprecated Use {@link net.minecraftforge.event.entity.player.AdvancementEvent.AdvancementEarnEvent} and {@link net.minecraftforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent} instead
  */
 public class AdvancementEvent extends PlayerEvent
```
The deprecated tag in the javadoc should be `@deprecated` in all lowercase. The annotation should also be present on the event, though not deprecating the event for removal:
```suggestion
 * @deprecated Use {@link net.minecraftforge.event.entity.player.AdvancementEvent.AdvancementEarnEvent} and {@link net.minecraftforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent} instead.
 */
@Deprecated(since = "1.19.2")
public class AdvancementEvent extends PlayerEvent
```

sciwhiz12 commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1011,6 +1011,10 @@ public static CriticalHitEvent getCriticalHit(Player player, Entity target, bool
+    /**
+     * @deprecated See {@link ForgeEventFactory#onAdvancementEarnedEvent} and {@link ForgeEventFactory#onAdvancementProgressedEvent}
+     */
+    //todo: remove and use ForgeEventFactory.onAdvancementEarned instead in 1.20
```
Please replace the TODO with a deprecation annotation marking it for removal. As we already patch in the new event factory to the same place, this only needs to be removed.
```suggestion
    @Deprecated(forRemoval = true, since = "1.19.2")
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/ForgeEventFactory.java:
```diff
@@ -820,4 +826,14 @@ public static WeightedRandomList<MobSpawnSettings.SpawnerData> getPotentialSpawn
+        MinecraftForge.EVENT_BUS.post(new AdvancementEarnEvent(player, earned));
+    }
+
+    public static void onAdvancementProgressedEvent(Player player, Advancement progressed, AdvancementProgress advancementProgress, String criterion, ProgressType progressType)
```
As we are moving towards making our public API more defined for consumers, please annotate these methods with `@ApiStatus.Internal`.

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -32,4 +34,99 @@ public Advancement getAdvancement()
+     * This event does not have a result. {@link HasResult}<br>
+     * <br>
+     * This event is fired on the {@link MinecraftForge#EVENT_BUS}.
+     */
```
Here's some cleaned up javadocs, using the (as of yet uncodified) format for event javadocs:
```suggestion
    /**
     * Fired when the player earns an advancement. An advancement is earned once its requirements are complete.
     *
     * <p>Note that advancements may be hidden from the player or used in background mechanics, such as recipe
     * advancements for unlocking recipes in the recipe book.</p>
     *
     * <p>This event is not {@linkplain Cancelable cancellable}, and does not {@linkplain HasResult have a result}.</p>
     *
     * <p>This event is fired on the {@linkplain MinecraftForge#EVENT_BUS main Forge event bus},
     * only on the {@linkplain LogicalSide#SERVER logical server}.</p>
     *
     * @see AdvancementProgress#isDone()
     */
```

sciwhiz12 commented on src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java:
```diff
@@ -32,4 +34,99 @@ public Advancement getAdvancement()
+     * This event does not have a result. {@link HasResult}<br>
+     * <br>
+     * This event is fired on the {@link MinecraftForge#EVENT_BUS}.
+     */
```
```suggestion
    /**
     * Fired when the player's progress on an advancement criterion is granted or revoked.
     *
     * <p>This event is not {@linkplain Cancelable cancellable}, and does not {@linkplain HasResult have a result}.</p>
     *
     * <p>This event is fired on the {@linkplain MinecraftForge#EVENT_BUS main Forge event bus},
     * only on the {@linkplain LogicalSide#SERVER logical server}.</p>
     *
     * @see AdvancementEarnEvent
     * @see net.minecraft.server.PlayerAdvancements#award(Advancement, String)
     * @see net.minecraft.server.PlayerAdvancements#revoke(Advancement, String)
     */
```

sciwhiz12 reviewed (changes requested):
I've requested some changes, at lot of it is just javadocs, formatting, and deprecation annotations for neatness. The advancement events themselves worked fine in testing.

---

We apologize if it seems that our reviews -- in particular, mine -- seemed to go back and forth. I requested changes as was befitting my information and the situation at the time, and changed them as I learned of new information and as the situation (such as impending RBs, which mean backwards compatibility) changed. 

In particular, modifying and extending an existing event becomes progressively more difficult as weeks pass by since an MC version's release (whether major or minor), because we have to consider the backwards compatibility of mods which rely on current behavior.

We also lack sufficient manpower and time to continually review all PRs in a timely fashion, which has resulted in some PRs being neglected for periods of time, which means the situation may have changed and new changes are required.

For example, my [review on Jan 9](https://github.com/MinecraftForge/MinecraftForge/pull/8360#pullrequestreview-847260103) was done near the release of 1.18.1, where breaking changes to the API was still allowed. My [subsequent review on Mar 27](https://github.com/MinecraftForge/MinecraftForge/pull/8360#pullrequestreview-922395089) was done close to the RB deadline for 1.18.2, which meant that breaking changes were _not_ allowed, and thus required changes to be backwards compatible.

As you've noticed and accordingly changed in your PR, we now have to maintain backwards compatibility with current 1.19 mods and mark breaking changes for 1.20. My immediate past review had already laid out the requirements to preserve backwards compatibility, so most of my requested changes here are to improve the readability and fix the proper annotations.

Again, we apologize if this has been troublesome. We are trying to improve the situation, as more of our team members become active in triaging once more. If you have more concerns or questions, you can contact me at the Forge discord as `sciwhiz12#1286`.

28Pollux28 committed:
implement changes suggested by @sciwhiz12

28Pollux28 commented:
That should be it. Sorry for my previous outburst, it was nor the place nor appropriate.

28Pollux28 committed:
removed unnecessary line in AdvancementEvent

TheCurle commented:
I'd appreciate if you add some further documentation on how these events work, on either [the official documentation](https://github.com/MinecraftForge/Documentation) or [the Community Wiki ](https://forge.gemwire.uk).
